_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+
+### Log for Computers First
+
+We'll never really stop reading logs, but it can become a chore to read 1000s of lines of logs, or look to see if multiple sets of logs from multiple devices may be experiencing the same issue. One of the best ways to get out of the business of reading logs, is to have our computers read it for us, and display it in a more digestible format.
+
+At my past companies, we developed shared command line tools that would read log files and print out a text summary of all the exceptions, any network connectivity issues detected, and various other useful information. The benefit of this was massive; not only did the tool analyze logs much faster than we could individually, but having a shared solution meant everyone benefited from the time savings, we all agreed upon the interpretation of different log messages, and we could add new features to the tool as the team grew and new scenarios were discovered.
+
+To facilitate that more easily, it would’ve helped if we could rely on the format of logs to be much more machine readable.
+
+#### Avoid Multiline Logs
+
+Multiline logs are a particularly painful problem for computers to parse. Most system expect logs to all be on a single line, and tracking logs across multiple lines requires retaining state specific to that error message just to parse that error.
+
+```markdown
+E bort    : Network error: -55
+E bort    :     RSSI: 61
+E bort    :     SSID: "Home-Guest"
+```
+
+On one line, it’s slightly less visually readable, but vastly more easily machine parsable.
+
+```markdown
+E bort    : Network error: -55, RSSI: 61, SSID: "Home-Guest"
+```
+
+#### Structured Logs
+
+We can take another page out of server logging best practices [^structuredlogging], and use structured logging. Logs by default are unstructured strings. If we logged more often in a structured format that was easily machine parsable, we could greatly simplify parsing and organizing the data. This is especially useful when we want to eventually parse specific information in the log, otherwise we'd have to write a Regex to parse the log.
+
+```markdown
+E bort    : {"Network error": "-55", "RSSI": 61, "SSID": "Home-Guest"}
+```
+
+JSON is the most common format for structured logs, and can be written manually in code fairly quickly. It's also fairly readable compared to the unstructured log, too.
+
+#### Skip Log Reading -- Build Visualizations!
+
+The end state of logging is ideally to never read the raw logs ourselves. Once we can empower computers to read all sorts of logs, we can build the tools to parse and then visualize the information, and save ourselves the effort of piecing together all the information in our head.
+
+<p align="center">
+  <img width="850" src="{% img_url blog-device-logging/logging-wifi.png %}" alt="Wi-Fi connection status over time" />
+</p>
```

Comment:
not specific to this article but I hate how the states are ordered and the colors used. It makes no sense. I hope we can fix that in the future.